### PR TITLE
refactor: MsgUpdateCrosschainFlags into MsgEnableCCTXFlags, MsgDisableCCTXFlags and MsgUpdateGasPriceIncreaseFlags

### PR DESCRIPTION
**Pull request by kingpinXD**
# Description

Refactor MsgUpdateCrosschainFlags to break it down into the following messages which are controlled by the mentioned admin polices 

```
MsgEnableCCTXFlags: PolicyType_groupOperational
MsgDisableCCTXFlags : PolicyType_groupEmegency
MsgUpdateGasPriceIncreaseFlags : PolicyType_groupOperational
```

Closes: https://github.com/zeta-chain/node/issues/2245

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Include instructions and any relevant details so others can reproduce. 

- [ ] Tested CCTX in localnet
- [ ] Tested in development environment
- [ ] Go unit tests
- [ ] Go integration tests
- [ ] Tested via GitHub Actions 

# Checklist:

- [ ] I have added unit tests that prove my fix feature works

### CODE REVIEW

File: x/observer/keeper/msg_server_update_gas_price_increase_flags.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+	if !isFound {
+		flags = *types.DefaultCrosschainFlags()
+		flags.IsInboundEnabled = false
+		flags.IsOutboundEnabled = false
+	}
+
+	err := msg.GasPriceIncreaseFlags.Validate()
+	if err != nil {
+		return &types.MsgUpdateGasPriceIncreaseFlagsResponse{}, err
+	}
```
Comment (kingpinXD) [inline x/observer/keeper/msg_server_update_gas_price_increase_flags.go:37]: Keeping here to retain existing functionality, but can be removed.

File: x/observer/keeper/msg_server_disable_cctx_flags_test.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+		_, err := srv.DisableCCTXFlags(sdk.WrapSDKContext(ctx), msg)
+		require.NoError(t, err)
+
+		flags, found := k.GetCrosschainFlags(ctx)
+		require.True(t, found)
+		require.False(t, flags.IsInboundEnabled)
+		require.False(t, flags.IsOutboundEnabled)
+	})
+
+	t.Run("can disable only 1 flag", func(t *testing.T) {
```
Comment (lumtis) [inline x/observer/keeper/msg_server_disable_cctx_flags_test.go:68]: We can split in two test: can disable inbound, and outbound

File: x/observer/keeper/msg_server_disable_cctx_flags_test.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+	"github.com/stretchr/testify/require"
+
+	keepertest "github.com/zeta-chain/zetacore/testutil/keeper"
+	"github.com/zeta-chain/zetacore/testutil/sample"
+	authoritytypes "github.com/zeta-chain/zetacore/x/authority/types"
+	"github.com/zeta-chain/zetacore/x/observer/keeper"
+	"github.com/zeta-chain/zetacore/x/observer/types"
+)
+
+func TestMsgServer_DisableCCTXFlags(t *testing.T) {
```
Comment (lumtis) [inline x/observer/keeper/msg_server_disable_cctx_flags_test.go:16]: We need to add assertion that the gas price increase flags are left untouched

File: x/observer/keeper/msg_server_enable_cctx_flags_test.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+	"github.com/stretchr/testify/require"
+
+	keepertest "github.com/zeta-chain/zetacore/testutil/keeper"
+	"github.com/zeta-chain/zetacore/testutil/sample"
+	authoritytypes "github.com/zeta-chain/zetacore/x/authority/types"
+	"github.com/zeta-chain/zetacore/x/observer/keeper"
+	"github.com/zeta-chain/zetacore/x/observer/types"
+)
+
+func TestMsgServer_EnableCCTXFlags(t *testing.T) {
```
Comment (lumtis) [inline x/observer/keeper/msg_server_enable_cctx_flags_test.go:16]: Same comments as for disable CCTXs

File: x/observer/keeper/msg_server_update_gas_price_increase_flags_test.go [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+	"github.com/stretchr/testify/require"
+
+	keepertest "github.com/zeta-chain/zetacore/testutil/keeper"
+	"github.com/zeta-chain/zetacore/testutil/sample"
+	authoritytypes "github.com/zeta-chain/zetacore/x/authority/types"
+	"github.com/zeta-chain/zetacore/x/observer/keeper"
+	"github.com/zeta-chain/zetacore/x/observer/types"
+)
+
+func TestKeeper_UpdateGasPriceIncreaseFlags(t *testing.T) {
```
Comment (lumtis) [inline x/observer/keeper/msg_server_update_gas_price_increase_flags_test.go:16]: Need to check CCTX enabled boolean are left untouched

File: x/observer/keeper/msg_server_update_gas_price_increase_flags_test.go [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+	t.Run("can update gas price increase flags if crosschain flags dont exist", func(t *testing.T) {
```
Comment (lumtis) [inline x/observer/keeper/msg_server_update_gas_price_increase_flags_test.go:17]: Need to add a test when flags already exist

File: x/observer/types/message_disable_cctx_flags.go [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+	}
+	return []sdk.AccAddress{creator}
+}
+
+func (msg *MsgDisableCCTXFlags) GetSignBytes() []byte {
+	bz := ModuleCdc.MustMarshalJSON(msg)
+	return sdk.MustSortJSON(bz)
+}
+
+func (msg *MsgDisableCCTXFlags) ValidateBasic() error {
```
Comment (lumtis) [inline x/observer/types/message_disable_cctx_flags.go:44]: We can add that at least one field must be `true`

File: x/observer/types/message_enable_cctx_flags.go [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+	}
+	return []sdk.AccAddress{creator}
+}
+
+func (msg *MsgEnableCCTXFlags) GetSignBytes() []byte {
+	bz := ModuleCdc.MustMarshalJSON(msg)
+	return sdk.MustSortJSON(bz)
+}
+
+func (msg *MsgEnableCCTXFlags) ValidateBasic() error {
```
Comment (lumtis) [inline x/observer/types/message_enable_cctx_flags.go:44]: Same as above

**Review by lumtis — COMMENTED** [review]
I propose we rename:
`MsgDisableCCTXFlags` -> `MsgPauseCCTX`
`MsgEnableCCTXFlags` -> `MsgUnpauseCCTX`

To keep consistency with PauseZRC20 for example.
Then we can use the pause/unpause terminology for all messages pausing functionalities in emergency

**Comment by kingpinXD** [comment]
> MsgPauseCCTX

I dont mind changing it but note 
- The zrc20 pause and unpause affects the paused value , for the zrc20 data structure directly , whereas here , we do not touch the CCTX struct at all .
- This would also make it inconsistent with `MsgUpdateGasPriceIncreaseFlags` even though they all modify the same data structure . 
- The values it affects are called `IsInboundEnabled` and `IsOutboundEnabled` and not `IsInboundPaused` and `IsOutboundPaused`.

**Comment by lumtis** [comment]
> > MsgPauseCCTX
> 
> I dont mind changing it but note
> 
> * The zrc20 pause and unpause affects the paused value , for the zrc20 data structure directly , whereas here , we do not touch the CCTX struct at all .
> * This would also make it inconsistent with `MsgUpdateGasPriceIncreaseFlags` even though they all modify the same data structure .
> * The values it affects are called `IsInboundEnabled` and `IsOutboundEnabled` and not `IsInboundPaused` and `IsOutboundPaused`.

Ok, let's keep it as is then

**Comment by kingpinXD** [comment]
> MsgEnableCCTX
> MsgDisableCCTX

The two values are under the `CrossChainFlags` struct, so verbally, does it make sense to treat them as flags?

**Comment by lumtis** [comment]
> > MsgEnableCCTX
> > MsgDisableCCTX
> 
> The two values are under the `CrossChainFlags` struct, so verbally, does it make sense to treat them as flags?

We will remove `CrossChainFlags` in the future since this structure pertain to the crosschain module.

And the enable/disable action is not performed to the flags but the CCTXs, flags is the structure that hold the information

File: docs/releases/v17_breaking_changes.md [hunk 9]
```diff
@@ -6,6 +6,10 @@
 * `MsgUpdateVerificationFlags` has been removed, and replaced with `MsgEnableHeaderVerification` and `MsgDisableHeaderVerification` messages.
     * `MsgEnableHeaderVerification` message enables block header verification for a list of chains and can be triggered via `PolicyType_groupOperational`
     * `MsgDisableHeaderVerification` message disables block header verification for a list of chains and can be triggered via `PolicyType_emergency`
+* `MsgUpdateCrosschainFlags` has been removed,and replaced with `MsgEnableCCTX`, `MsgDisableCCTX` and `MsgUpdateGasPriceIncreaseFlags` messages.
```
Comment (lumtis) [inline docs/releases/v17_breaking_changes.md:9]:
```suggestion
* `MsgUpdateCrosschainFlags` has been removed, and replaced with `MsgEnableCCTX`, `MsgDisableCCTX` and `MsgUpdateGasPriceIncreaseFlags` messages.
```

**Review by lumtis — APPROVED** [review]

File: docs/releases/v17_breaking_changes.md (hunk 9 above)
Comment (lumtis) [inline docs/releases/v17_breaking_changes.md:9]: Actually we also need to move this to a new section other than `Verification Flags update`

**Review by lumtis — APPROVED** [review]

**Pull request merged**